pkg/ephemeral/ephemeral.go

```diff
@@ -0,0 +1,115 @@
+package ephemeral
+
+import (
+	"github.com/kanisterio/kanister/pkg/kube"
+	corev1 "k8s.io/api/core/v1"
+)
+
+var (
+	Container  ApplierList[corev1.Container]
+	PodOptions ApplierList[kube.PodOptions]
+)
+
+// ApplierSet is a group of Appliers, typically returned by an constructor.
+type ApplierSet struct {
+	Container  Applier[corev1.Container]
+	PodOptions Applier[kube.PodOptions]
+}
+
+// Register generically registers an Applier.
+func Register[T Constraint](applier Applier[T]) {
+	switch applier := any(applier).(type) {
+	case Applier[corev1.Container]:
+		Container.Register(applier)
+	case Applier[kube.PodOptions]:
+		PodOptions.Register(applier)
+	default:
+		panic("Unknown applier type")
+	}
+}
+
+// RegisterSet registers each of the Appliers contained in the set.
+func RegisterSet(set ApplierSet) {
+	if set.Container != nil {
+		Container.Register(set.Container)
+	}
+
+	if set.PodOptions != nil {
+		PodOptions.Register(set.PodOptions)
+	}
+}
+
+// Constraint provides the set of types allowed for appliers and filterers.
+type Constraint interface {
+	kube.PodOptions | corev1.Container
+}
+
+// Applier is the interface which applies a manipulation to the PodOption to be
+// used to run ephemeral pdos.
+type Applier[T Constraint] interface {
+	Apply(*T)
+}
+
+// ApplierFunc is a function which implements the Applier interface and can be
+// used to generically manipulate the PodOptions.
+type ApplierFunc[T Constraint] func(*T)
+
+func (f ApplierFunc[T]) Apply(options *T) { f(options) }
+
+// ApplierList is an array of registered Appliers which will be applied on
+// a PodOption.
+type ApplierList[T Constraint] []Applier[T]
+
+// Apply calls the Applier::Apply method on all registered appliers.
+func (l ApplierList[T]) Apply(options *T) {
+	for _, applier := range l {
+		applier.Apply(options)
+	}
+}
+
+// Register adds the applier to the list of Appliers to be used when
+// manipulating the PodOptions.
+func (l *ApplierList[T]) Register(applier Applier[T]) {
+	*l = append(*l, applier)
+}
+
+// Filterer is the interface which filters the use of registered appliers to
+// only those PodOptions that match the filter criteria.
+type Filterer[T Constraint] interface {
+	Filter(*T) bool
+}
+
+// FiltererFunc is a function which implements the Filterer interface and can be
+// used to generically filter PodOptions to manipulate using the ApplierList.
+type FiltererFunc[T Constraint] func(*T) bool
+
+func (f FiltererFunc[T]) Filter(options *T) bool {
+	return f(options)
+}
+
+// Filter applies the Applier's if the Filterer criterion is met.
+func Filter[T Constraint](filterer Filterer[T], appliers ...Applier[T]) Applier[T] {
+	return ApplierFunc[T](func(options *T) {
+		if !filterer.Filter(options) {
+			return
+		}
+
+		for _, applier := range appliers {
+			applier.Apply(options)
+		}
+	})
+}
+
+// PodOptionsNameFilter is a Filterer that filters based on the PodOptions.Name.
```

Comment:
`PodOptions.Name` is the pod name. Wonder if we should explicitly state that in the comment

Reply:
Sounds good. I'll add more to the comment.